src/dag.ts

```diff
@@ -68,25 +71,27 @@ export function iterate<T>(nodes: INode<T>[], edges: IEdge[],
     return unvisited === false ? childNodes : childNodes.filter(filterNotVisited);
   }
 
-  let i = 0;
-  nodeStack.push(getFirstNode());
+  Array.prototype.push.apply(nodeStack, getFirstNodes());
 
   while (nodeStack.length) {
     // Take a layer from the stack
     const node = nodeStack.pop();
 
-    // Mark this node as visited
-    nodeVisited[node] = true;
-
     // Collect the previous Layers
     const parentNodes = getParentNodes(node);
 
+    // Mark this node as visited and record its layer count as
+    // max(parents) + 1
+    nodeVisited[node] = parentNodes.reduce(function (accu, curr) {
+      return (accu > nodeVisited[curr] + 1) ? accu : nodeVisited[curr] + 1;
+    }, 0);
+
     // Get the layer and previous layers
     const layer = getLayerFromNode(node);
     const parentLayers = parentNodes.map(getLayerFromNode);
     
     // Call the iterator callback
-    iteratorFn.call(null, layer, parentLayers, i++);
+    iteratorFn.call(null, layer, parentLayers, nodeVisited[node]);
```

Comment:
do you need the layer count in the iterator function for something specific?

Reply:
The layer index is useful for drawing the graph.